src/base/http/connection.h

```diff
@@ -27,41 +28,69 @@
  * exception statement from your version.
  */
 
-
 #pragma once
 
 #include <QElapsedTimer>
 #include <QObject>
 
+#include "types.h"
+
 class QTcpSocket;
 
 namespace Http
 {
-    class IRequestHandler;
-    struct Response;
-
     class Connection : public QObject
     {
         Q_OBJECT
         Q_DISABLE_COPY(Connection)
 
     public:
-        Connection(QTcpSocket *socket, IRequestHandler *requestHandler, QObject *parent = nullptr);
-        ~Connection();
+        enum Status
+        {
+            PendingRequest,
+            PendingStatus,
+            PendingHeaders,
+            PendingContent,
+            WritingContent,
+            Closed
+        };
+
+        // Connection takes ownership of socket
+        explicit Connection(QTcpSocket *socket, QObject *parent = nullptr);
 
-        bool hasExpired(qint64 timeout) const;
-        bool isClosed() const;
+        void sendStatus(const ResponseStatus &status);
+        void sendHeaders(const HeaderMap &headers);
+        void sendContent(const QByteArray &data);
```

Comment:
You didn't really send out anything by calling them, `add***` would be perhaps more correct.

Reply:
Can't we see it's not sending content immediately as an implementation detail. I mean Python's HTTP server also has the same functions and they too don't send the data immediately. `add***` in Connection doesn't seem right.

Reply:
Let me explain what I would expect:
```c++
Response res;  // a specific class for building response 
res.setStatus(...);
res.setHeaders(...);
res.setContent(...);

connection.send(res);  // really send it out from socket
```
The current way of `send*` functions are not intuitive for general use and couldn't be more wrong as you are mixing higher layer operations (http) with low level socket operations.

Reply:
IIRC, his main reason is the need to send very large data (several gigabytes) in a single HTTP response.

Reply:
I aimed to implement something like Python Http Handler https://docs.python.org/3/library/http.server.html#http.server.BaseHTTPRequestHandler.send_header 
it's used by a large audience so IMO there is nothing wrong with this PR's abstractions. I am open to ideas.

Reply:
>IIRC, his main reason is the need to send very large data (several gigabytes) in a single HTTP response.

Without context about the problem aiming to solve, it is impossible to envision the proper solution. However, the task should be quite easy in the previous example:
```c++
Response res;
// prepare the start of the response:
res.setStatus(...);
res.setHeaders(...);
connection.send(res);  // send response headers without the content

// ...

connection.send(...);  // send whatever data/bytes
// or if the content size is reasonable
Response res;
res.setContent(...);
connection.send(res);  // send content without other headers/status

// The connection::send() only cares about sending *bytes* out and not about higher layer things
```

>it's used by a large audience 

Seriously? Because it is easy to start up doesn't mean it is well designed, there is this warning at the top of its [documentation](https://docs.python.org/3/library/http.server.html):
>Warning http.server is not recommended for production. It only implements basic security checks.

Also someone views it as a toy implementation at best: https://stackoverflow.com/a/26566915

>I am open to ideas.

There are so many web/network framework out there and you fav one that is questionable. If I were to pick, at least boost asio & boost beast wouldn't partition the layers wrong and golang official http package looks nice too.

Reply:
> Without context about the problem aiming to solve, it is impossible to envision the proper solution

the idea is to use this functionality to implement HTTP range requests (https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) in an asynchronous fashion which is not possible with the current master's IRequestHandler

>
>Response res;
>res.setContent(...);
>connection.send(res);  // send content without other headers/status

wouldn't this break the basic structure of HTTP response?

> There are so many web/network framework out there and you fav one that is questionable. If I were to pick, at least boost asio & boost beast wouldn't partition the layers wrong and golang official http package looks nice too.

I never used any of those. 
IMO the best thing I can do is to move the response serializing into a separate class but by that logic, we should have a separate class reading and serializing the request from the underlying socket as well but then what would the Connection class do.

Reply:
It is a very good idea to implement multi-layer logic (this applies not only to the network stack).
What do we have here? We use the TCP socket (provided by Qt) and implement the HTTP layer on top of it. And this new layer should provide an interface corresponding to HTTP. I.e. it reads data from the socket and parses it according to the HTTP protocol into some structured representation (HTTP Request), which it then provides to its clients (the caller side). And back from the clients (from the caller side) it should accept the HTTP Response in a protocol-structured form, then serialize it and send it to the underlying TCP layer.
In the base case (which nevertheless covers most use cases) it should accept the Response object that contains the complete response data (status, headers, content), check it for consistency, serialize it entirely, and write to the socket. So such an interface (accepting Response) should definitely exist.

If HTTP really allows sending such large data, then the approach above can be inapplicable, since we need to be able to send data sequentially, for example, reading it from a file (etc.). But we still need to take care not to go beyond the HTTP protocol.
I imagine that this can be implemented through some more advanced version of the Response, an instance of which will be created by the factory method of Connection (since it must be assigned a socket that shouldn't be publicly accessible), immediately receiving the Status and Headers as input parameters, and then it will act as a stream, receiving data and writing it to the socket (and checking written data size against content length specified when it is created).

Reply:
> I imagine that this can be implemented through some more advanced version of the Response, an instance of which will be created by the factory method of Connection (since it must be assigned a socket that shouldn't be publicly accessible), immediately receiving the Status and Headers as input parameters, and then it will act as a stream, receiving data and writing it to the socket (and checking written data size against content length specified when it is created).

maybe what you meant can be done by making Response::content a QIODevice (https://github.com/qbittorrent/qBittorrent/blob/master/src/base/http/types.h#L121). I tried it once but this fairly complicates other parts of the torrent file implementation. We're not implementing a library here and our use cases are limited. IMO the current implementation is small and simple enough to not cause any problem or maintainer burden in the future.

Reply:
E.g.:
```c++
std::shared_ptr<StreamResponse> resp = connection->beginStreamResponse(status, headers, contentLength);
resp->write(data); // repeat as many times as you want until the content length is reached
```
StreamResponse instance should become invalid after content length is reached and throw an exception if you try to write more data.
Connection should disallow another Responses when any StreamResponse is active.

Reply:
>I tried it once but this fairly complicates other parts of the torrent file implementation.

The example above doesn't look any more complicated than what you're suggesting in this PR (in terms of the interface), even though it's defined in terms of HTTP.

Reply:
Let's also wait for @Chocobo1 opinion about my latest comments.

Reply:
To clarify once again (if this was unclear from my previous comments), I suggest having two types of Responses, one for general cases and the other for the case of sequential write content.

Reply:
> The example above doesn't look any more complicated than what you're suggesting in this PR (in terms of the interface), even though it's defined in terms of HTTP.

HTTP responses are well defined. your solution creates two branches in the behavior of a concrete class. I read somewhere that this is not a good thing. may be related to single responsibility principal.

Reply:
>To clarify once again (if this was unclear from my previous comments), I suggest having two types of Responses, one for general cases and the other for the case of sequential write content.

Agree, I think your suggestion really make the interface robust enough to avoid (most) misuse and would still preserve the correct layering.

>https://github.com/qbittorrent/qBittorrent/pull/14234#discussion_r567232574

In the simplified example, I want to note that it might need checking if the socket is still open before sending data.

Reply:
>may be related to single responsibility principal.

If you talk about Http::Connection it is still responsible for sending/receiving HTTP "packets" (requests/responses).
>your solution creates two branches in the behavior of a concrete class. 

In fact, you can use the second one (https://github.com/qbittorrent/qBittorrent/pull/14234#discussion_r567232574) in all cases.
>HTTP responses are well defined.

Yes.
I agree that it is possible to have the simplest interface with three methods (as you suggest) and make the correctness of their use the responsibility of the user of the class. And this should not even cause any problems in the simplest case, when you call them in a row in one place. But how to control the correctness of the logic in more complex cases (for example, in what you intend to implement next), where the calls of these methods are located far from each other (both in code and at runtime)? How are you going to check the correctness of your logic? Just inserting assertions and waiting for them to hit? Assertions is a good tool that should not be neglected, because we often can not predict all the mistakes. But why not add to it an interface that eliminates the possibility of making such mistakes (at least some of the possible ones)?

Reply:
ok now it make sense.

How about this - 

```c++
class ResponseContent
{
public:
 virtual size_t size() const = 0;
 virtual size_t read(void *data, size_t len) = 0;
signals:
 void readyRead();
};
```

and provide an implementation based on QByteArray for convenience;

Reply:
>How about this -

I'm sorry, it doesn't make sense to me until you show me how you're going to use it.

Reply:
make Response::Content (https://github.com/qbittorrent/qBittorrent/blob/master/src/base/http/types.h#L12) a std::shared_ptr<ResponseContent>
then everything will remain as it is in the current master. I just need to change `Connection::sendResponse` to handle this async type operation.